src/core/externalstorage/qgsexternalstorage.h

```diff
@@ -0,0 +1,199 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSEXTERNALSTORAGE_H
+#define QGSEXTERNALSTORAGE_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+
+#include <QObject>
+#include <QString>
+
+class QgsExternalStorageFetchedContent;
+class QgsExternalStorageStoredContent;
+
+/**
+ * \ingroup core
+ * \brief Abstract interface for external storage - to be implemented by various backends
+ * and registered in QgsExternalStorageRegistry.
+ *
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsExternalStorage
+{
+  public:
+
+    /**
+     * Destructor
+     */
+    virtual ~QgsExternalStorage() = default;
+
+    /**
+     * Unique identifier of the external storage type.
+     */
+    virtual QString type() const = 0;
+
+    /**
+     * Store file \a filePath to the \a url for this project external storage.
+     * Storing process is run in background.
+     * Returns a QgsExternalStorageStoredContent to follow the status of the stored resource.
+     *
+     * After using this method, user should check if the returned content is not already finished
+     * (storing could be instantaneous, if file has already been stored for instance) and then wait
+     * for QgsExternalStorageStoredContent::stored(), QgsExternalStorageStoredContent::errorOccurred() or
+     * QgsExternalStorageStoredContent::canceled() signals.
+     *
+     * It's possible to give \a authcfg authentification configuration id in case its needed.
+     *
+     * Caller takes ownership of the returned symbol.
+     */
+    virtual QgsExternalStorageStoredContent *store( const QString &filePath, const QString &url, const QString &authcfg = QString() ) const = 0 SIP_FACTORY;
+
+    /**
+     * Fetch file from \a url for this project external storage.
+     * Fetching process is run in background.
+     * Returns a QgsExternalStorageFetchedContent to follow the status of the fetched resource.
+     *
+     * After using this method, user should check if the returned content is not already finished
+     * (fetching could be instantaneous, if file has already been fetched and cached for instance)
+     * and then wait for QgsExternalStorageStoredContent::fetched(), QgsExternalStorageStoredContent::errorOccurred() or
+     * QgsExternalStorageStoredContent::canceled() signals.
+     *
+     * It's possible to give \a authcfg authentification configuration id in case its needed.
+     */
+    virtual QgsExternalStorageFetchedContent *fetch( const QString &url, const QString &authcfg = QString() ) const = 0 SIP_FACTORY;
+};
+
+/**
+ * \ingroup core
+ * \brief Base class for QgsExternalStorage stored and fetched content
+ *
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsExternalStorageOperation : public QObject
+{
+    Q_OBJECT
+
+  public:
+
+    //! Status of fetched content
+    enum ContentStatus
```

Comment:
Please move this to the Qgis class and promote to an enum class. We're moving toward having all core enums in the Qgis class (Qt style) to facilitate forward declaration vs header imports.

Reply:
Done